From d858cf5c198e648d270f5e8a945a6974c05e3404 Mon Sep 17 00:00:00 2001
From: Damien Thenot <damien.thenot@vates.tech>
Date: Thu, 28 May 2026 16:15:22 +0200
Subject: [PATCH] fix(cleanup): fix for live leaf coalesce
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

In case of two leaf coalesce following each other, tapdisk would crash.
```
*b43ce641[qcow2](1.000G///1.004G|ao)
    *e735e552[qcow2](1.000G///1.004G|ao)
        340334d1[qcow2](1.000G///1.004G|ao)
```
340334d1 would be merged in e735e552 following a commit.
Then e735e552 would be renamed in 340334d1 to maintain the same UUID for the leaf.
tapdisk would automatically be using e735e552 as leaf after the commit but
would not know about the rename. The next leaf coalesce would refer to the leaf
with the same name but tapdisk would not know to refer to e735e552 instead. We
introduced a refresh of the VDI so tapdisk read updated information from the
chain after each leaf coalesce.

Signed-off-by: Damien Thenot <damien.thenot@vates.tech>
---
 drivers/cleanup.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/cleanup.py b/drivers/cleanup.py
index bc2ba837..d559efd1 100755
--- a/drivers/cleanup.py
+++ b/drivers/cleanup.py
@@ -2905,6 +2905,10 @@ class SR(object):
             _, host_ref = next(iter(host_refs.items()))
             vdi._coalesceCowImageOnHost(host_ref, vdi) # vdi is the leaf for the online coalesce
             util.fistpoint.activate("LVHDRT_coaleaf_after_coalesce", self.uuid)
+            vdi.pause(failfast=True)
+            # We make a pause here after the online coalesce but before the rename so we can refresh the chain for tapdisk. 
+            # It's also needed to be paused for the rename on slaves with LVMSR.
+            # We let the caller `_liveLeafCoalesce` do the unpause with the call to `vdi.ensureUnpaused()`
         else:
             vdi.validate(True)
             vdi.parent.validate(True)
